File: hearth/data/store.py
"""DataStore: the seam between every loop and the database.

Methods mirror butterbase's auto-REST CRUD semantics (insert / get /
query-with-filters-sort-limit / update / delete) so that swapping the
SQLite-backed implementation for a ButterbaseStore later requires no
changes to callers.
"""
from abc import ABC, abstractmethod
from typing import Any, Optional

Record = dict[str, Any]
Filters = dict[str, Any]
Sort = list[tuple[str, str]]


class DataStore(ABC):
    @abstractmethod
    def insert(self, table: str, record: Record) -> Record:
        """Insert a record, returning it with its assigned id."""

    @abstractmethod
    def get(self, table: str, id: int) -> Optional[Record]:
        """Fetch a single record by id, or None."""

    @abstractmethod
    def query(
        self,
        table: str,
        filters: Optional[Filters] = None,
        sort: Optional[Sort] = None,
        limit: Optional[int] = None,
    ) -> list[Record]:
        """Query records.

        filters: {column: value} for equality, or {column: {"gte"|"lte"|"gt"|"lt"|"ne": value}}
        sort: [(column, "asc"|"desc"), ...]
        """

    @abstractmethod
    def update(self, table: str, id: int, patch: Record) -> Record:
        """Patch a record by id, returning the updated record."""

    @abstractmethod
    def delete(self, table: str, id: int) -> None:
        """Delete a record by id."""

File: src/data/store.ts
// DataStore mirrors butterbase.ai's REST CRUD semantics (insert / get /
// query-with-filters-sort-limit / update / delete). Every loop depends only on
// this interface, never on SQLite directly — see sqliteStore.ts for the swap
// comment explaining how a ButterbaseStore drops in later.

export interface RangeFilter {
  gte?: unknown;
  lte?: unknown;
  gt?: unknown;
  lt?: unknown;
}

export interface QueryOptions {
  filters?: Record<string, unknown | RangeFilter>;
  sort?: Array<[string, "asc" | "desc"]>;
  limit?: number;
}

export interface DataStore {
  insert<T>(table: string, record: Omit<T, "id">): Promise<T>;
  get<T>(table: string, id: number): Promise<T | null>;
  query<T>(table: string, options?: QueryOptions): Promise<T[]>;
  update<T>(table: string, id: number, patch: Partial<T>): Promise<T>;
  delete(table: string, id: number): Promise<void>;
}
